cmd/tiller/release_server.go

```diff
@@ -737,22 +742,53 @@ func (s *releaseServer) performRelease(r *release.Release, req *services.Install
 		}
 	}
 
-	// regular manifests
-	kubeCli := s.env.KubeClient
-	b := bytes.NewBufferString(r.Manifest)
-	if err := kubeCli.Create(r.Namespace, b); err != nil {
-		log.Printf("warning: Release %q failed: %s", r.Name, err)
-		r.Info.Status.Code = release.Status_FAILED
-		s.recordRelease(r, req.ReuseName)
-		return res, fmt.Errorf("release %s failed: %s", r.Name, err)
+	// if this is a replace operation, append to the release history
+	if req.ReuseName {
+		switch h, err := s.env.Releases.History(req.Name); {
+		case err == nil && len(h) >= 1:
+			// get latest release revision
+			relutil.Reverse(h, relutil.SortByRevision)
+
+			// old release
+			old := h[0]
+
+			// update old release status
+			old.Info.Status.Code = release.Status_SUPERSEDED
+			s.recordRelease(old, true)
+
+			// update new release with next revision number
+			// so as to append to the old release's history
+			r.Version = old.Version + 1
+
+			if err := s.performKubeUpdate(old, r); err != nil {
+				log.Printf("warning: Release replace %q failed: %s", r.Name, err)
+				old.Info.Status.Code = release.Status_SUPERSEDED
+				r.Info.Status.Code = release.Status_FAILED
+				s.recordRelease(old, true)
+				s.recordRelease(r, false)
+				return res, err
+			}
+
+		default:
+			// nothing to replace
```

Comment:
In the default case, (i.e. no previous release, but 'replace' flag was provided), the resource should still be created as in the else block.